### PR TITLE
[RLlib; Offline RL] Add support to directly read from episodes.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

The Offline RL API in the new stack will soon support recording capabilities (see #46818) that allow direclty storing `RLlib`'s episodes (i.e. `SingleAgentEpisode`s). In regard to this upcoming change this PR proposes a direct episode readfing option for the `OfflinePreLearner`  that can skip the `_map_to_episodes` step in the `map_batches` dta pipeline and therefore increase training speed.

## Related issue number

Relates to #46818 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/marwil/tests/test_marwil_old.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
+import numpy as np
```
Comment (sven1977) [inline rllib/algorithms/marwil/tests/test_marwil_old.py:1]: nit: rename into "...old_api_stack.py" for more clarity.

File: rllib/BUILD [hunk 2]
```diff
@@ -909,7 +937,7 @@ py_test(
 
 # MARWIL
 py_test(
-    name = "test_marwil",
+    name = "test_marwil_old",
```
Comment (sven1977) [inline rllib/BUILD:940]: same: rename into "...old_api_stack" for clarity.

File: rllib/algorithms/marwil/marwil_offline_prelearner.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                del batch.policy_batches[module_id]
+
+        # TODO (simon): Log steps trained for metrics (how?). At best in learner
+        # and not here. But we could precompute metrics here and pass it to the learner
+        # for logging. Like this we do not have to pass around episode lists.
+
+        # TODO (simon): episodes are only needed for logging here.
+        return {"batch": [batch]}
+
+    def _compute_gae_from_episodes(
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil_offline_prelearner.py:62]: Any chance we can share this code with PPO's?

Create a utility function in PPO folder (e.g. algorithms/ppo/utils.py), then import and use this same function for MARWIL.

File: rllib/algorithms/marwil/marwil.py [hunk 4]
```diff
@@ -92,6 +113,7 @@ def __init__(self, algo_class=None):
         self.input_ = "sampler"
         self.postprocess_inputs = True
         self.lr = 1e-4
+        self.lambda_ = 1.0
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:116]: Was this parameter missing before?

File: rllib/algorithms/marwil/marwil.py [hunk 5]
```diff
@@ -157,6 +179,36 @@ def training(
...
+            )
+
+            return SingleAgentRLModuleSpec(
+                module_class=MARWILTorchRLModule,
+                catalog_class=MARWILCatalog,
+            )
+        else:
+            raise ValueError(
+                f"The framework {self.framework_str} is not supported. "
+                "Use either 'torch' or 'tf2'."
```
Comment (sven1977) [inline rllib/algorithms/marwil/marwil.py:196]: nit: "Use 'torch' instead." :)

File: rllib/algorithms/marwil/marwil_offline_prelearner.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil_offline_prelearner.py:62]: That's the question I have: DO we really need the complete code? Or can we get rid of either the batch or episodes in their when considering MARWIL and the offline learning pipeline?

File: rllib/algorithms/marwil/marwil.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil.py:116]: Appeared to - I got errors until I added this parameter in its config. This should normally be a default parameter all algorithms have, but somehow it gets lost on the way down.

File: rllib/algorithms/marwil/marwil.py [hunk 6]
```diff
@@ -248,6 +320,87 @@ def get_default_policy_class(
 
     @override(Algorithm)
     def training_step(self) -> ResultDict:
+        if self.config.enable_env_runner_and_connector_v2:
+            return self._training_step_new_stack()
+        elif self.config.enable_rl_module_and_learner:
+            raise ValueError(
+                "`enable_rl_module_and_learner=True`. Hybrid stack is not "
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil.py:327]: Great catch!

File: rllib/algorithms/marwil/marwil.py [hunk 7]
```diff
@@ -281,3 +434,88 @@ def training_step(self) -> ResultDict:
         self.env_runner.set_global_vars(global_vars)
 
         return train_results
+
+    def _training_step_hybrid_step(self) -> ResultDict:
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil.py:438]: Another good catch! We should go fishing

File: rllib/algorithms/marwil/marwil_catalog.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+from ray.rllib.core.models.base import ActorCriticEncoder, Encoder, Model
+from ray.rllib.utils.annotations import override, OverrideToImplementCustomLogic
+
+
+class MARWILCatalog(Catalog):
+    """The Catalog class used to build models for MARWIL.
+
+    MARWILCatalog provides the following models:
+
+
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil_catalog.py:20]: Yeah missed that. Godd catch Sven!

File: rllib/algorithms/marwil/marwil.py (hunk 4 above)
Comment (sven1977) [reply rllib/algorithms/marwil/marwil.py:115]: :+1:
I don't think this should be a default param, b/c algos like DQN (or even IMPALA) don't need it. Only the GAE-using ones.

Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil.py:115]: Ah okay, then it makes sense.

**Review by sven1977 — APPROVED** [review]
LGTM now! Thanks for this PR @simonsays1980 !

**Pull request merged**